wolfcrypt/src/ascon.c

```diff
@@ -0,0 +1,528 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+#ifdef HAVE_CONFIG_H
+    #include <config.h>
+#endif
+
+#include <wolfssl/wolfcrypt/settings.h>
+
+#ifdef HAVE_ASCON
+
+#include <wolfssl/wolfcrypt/ascon.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+#include <wolfssl/wolfcrypt/logging.h>
+#ifdef NO_INLINE
+    #include <wolfssl/wolfcrypt/misc.h>
+#else
+    #define WOLFSSL_MISC_INCLUDED
+    #include <wolfcrypt/src/misc.c>
+#endif
+
+/*
+ * Implementation of the ASCON AEAD and HASH algorithms. Based on the NIST
+ * Initial Public Draft "NIST SP 800-232 ipd" and reference implementation found
+ * at https://github.com/ascon/ascon-c.
+ */
+
+/*
+ * TODO
+ * - Add support for big-endian systems
+ * - Add support for 32-bit and smaller systems */
+
+#define MAX_ROUNDS 12
+
+#ifndef WOLFSSL_ASCON_UNROLL
+
+/* Table 4 */
+static const byte round_constants[MAX_ROUNDS] = {
+    0xf0, 0xe1, 0xd2, 0xc3, 0xb4, 0xa5, 0x96, 0x87, 0x78, 0x69, 0x5a, 0x4b
+};
+
+static byte start_index(byte rounds)
+{
+    switch (rounds) {
+        case 6:
+            return 6;
+        case 8:
+            return 4;
+        case 12:
+            return 0;
+        default:
+            WOLFSSL_MSG("Something went wrong in wolfCrypt logic. Wrong ASCON "
+                        "rounds value.");
+            return MAX_ROUNDS;
+    }
+}
+
+static WC_INLINE void ascon_round(AsconState* a, byte round)
+{
+    AsconState tmp;
+    /* 2.6.1 Addition of Constants */
+    a->s64[2] ^= round_constants[round];
+    /* 2.6.2 Substitution Layer */
+    a->s64[0] ^= a->s64[4];
+    a->s64[4] ^= a->s64[3];
+    a->s64[2] ^= a->s64[1];
+    tmp.s64[0] = a->s64[0] ^ (~a->s64[1] & a->s64[2]);
+    tmp.s64[2] = a->s64[2] ^ (~a->s64[3] & a->s64[4]);
+    tmp.s64[4] = a->s64[4] ^ (~a->s64[0] & a->s64[1]);
+    tmp.s64[1] = a->s64[1] ^ (~a->s64[2] & a->s64[3]);
+    tmp.s64[3] = a->s64[3] ^ (~a->s64[4] & a->s64[0]);
+    tmp.s64[1] ^= tmp.s64[0];
+    tmp.s64[3] ^= tmp.s64[2];
+    tmp.s64[0] ^= tmp.s64[4];
+    tmp.s64[2] = ~tmp.s64[2];
+    /* 2.6.3 Linear Diffusion Layer */
+    a->s64[4] =
+         tmp.s64[4] ^ rotrFixed64(tmp.s64[4],  7) ^ rotrFixed64(tmp.s64[4], 41);
+    a->s64[1] =
+         tmp.s64[1] ^ rotrFixed64(tmp.s64[1], 61) ^ rotrFixed64(tmp.s64[1], 39);
+    a->s64[3] =
+         tmp.s64[3] ^ rotrFixed64(tmp.s64[3], 10) ^ rotrFixed64(tmp.s64[3], 17);
+    a->s64[0] =
+         tmp.s64[0] ^ rotrFixed64(tmp.s64[0], 19) ^ rotrFixed64(tmp.s64[0], 28);
+    a->s64[2] =
+         tmp.s64[2] ^ rotrFixed64(tmp.s64[2],  1) ^ rotrFixed64(tmp.s64[2],  6);
+}
+
+static void permutation(AsconState* a, byte rounds)
+{
+    byte i = start_index(rounds);
+    for (; i < MAX_ROUNDS; i++) {
+        ascon_round(a, i);
+    }
+}
+
+#else
+
+#define p(a, c) do {                                                           \
+    AsconState tmp;                                                            \
```

Comment:
Use individual word64 variables to help compiler optimize.

Reply:
Changing.